### PR TITLE
WIP: feat(prometheus): add prometheus endpoint providing basic request met…

### PR DESCRIPTION
**Pull request by ecktom**
## Proposed changes

This will add an basic prometheus endpoint providing request metrics to get some application insights

## Checklist

- [x] I have read the [contributing guidelines](../blob/master/CONTRIBUTING.md).
- [x] I have read the [security policy](../security/policy).
- [x] I confirm that this pull request does not address a security vulnerability. If this pull request addresses a security.
vulnerability, I confirm that I got green light (please contact [security@ory.sh](mailto:security@ory.sh)) from the maintainers to push the changes.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] I have added or changed [the documentation](docs/docs).

### CODE REVIEW

**Comment by aeneasr** [comment]
Sorry, I forgot to press on "submit review" :)

File: cmd/server/server.go [hunk 1]
```diff
@@ -189,18 +190,36 @@ func RunServe(version, build, date string) func(cmd *cobra.Command, args []strin
...
 			publicmw.Use(tracer)
 		}
 
+		prometheusRepo, err := metrics.NewPrometheusRepository(logger)
+		if err != nil {
+			logger.Warnf("cannot create proemtehus repo: %+v", err)
+		}
+
+		// TODO: Make prom settings (port/path) configurable
+		promAddr := ":9124"
```
[the remark this answers is not in the document]
Comment (ecktom) [reply cmd/server/server.go:207]: Ehm, what admin api endpoint? :D

**Comment by ecktom** [comment]
So I'm not sure what you mean regarding the config but I've build the version from the PR locally and tested it by curling the scraping endpoint which looks good:

```
/ # curl localhost:9124
# HELP go_gc_duration_seconds A summary of the pause duration of garbage collection cycles.
# TYPE go_gc_duration_seconds summary
go_gc_duration_seconds{quantile="0"} 6.787e-06
go_gc_duration_seconds{quantile="0.25"} 2.4558e-05
go_gc_duration_seconds{quantile="0.5"} 4.1088e-05
go_gc_duration_seconds{quantile="0.75"} 0.000107255
go_gc_duration_seconds{quantile="1"} 0.003268809
go_gc_duration_seconds_sum 0.006710868
go_gc_duration_seconds_count 37
# HELP go_goroutines Number of goroutines that currently exist.
# TYPE go_goroutines gauge
go_goroutines 27
# HELP go_info Information about the Go environment.
# TYPE go_info gauge
go_info{version="go1.14"} 1
# HELP go_memstats_alloc_bytes Number of bytes allocated and still in use.
# TYPE go_memstats_alloc_bytes gauge
go_memstats_alloc_bytes 4.355816e+06
# HELP go_memstats_alloc_bytes_total Total number of bytes allocated, even if freed.
# TYPE go_memstats_alloc_bytes_total counter
go_memstats_alloc_bytes_total 1.16957672e+08
# HELP go_memstats_buck_hash_sys_bytes Number of bytes used by the profiling bucket hash table.
# TYPE go_memstats_buck_hash_sys_bytes gauge
go_memstats_buck_hash_sys_bytes 1.466244e+06
# HELP go_memstats_frees_total Total number of frees.
# TYPE go_memstats_frees_total counter
go_memstats_frees_total 480030
# HELP go_memstats_gc_cpu_fraction The fraction of this program's available CPU time used by the GC since the program started.
# TYPE go_memstats_gc_cpu_fraction gauge
go_memstats_gc_cpu_fraction 0.00019696603142984264
# HELP go_memstats_gc_sys_bytes Number of bytes used for garbage collection system metadata.
# TYPE go_memstats_gc_sys_bytes gauge
go_memstats_gc_sys_bytes 3.582216e+06
# HELP go_memstats_heap_alloc_bytes Number of heap bytes allocated and still in use.
# TYPE go_memstats_heap_alloc_bytes gauge
go_memstats_heap_alloc_bytes 4.355816e+06
# HELP go_memstats_heap_idle_bytes Number of heap bytes waiting to be used.
# TYPE go_memstats_heap_idle_bytes gauge
go_memstats_heap_idle_bytes 5.8769408e+07
# HELP go_memstats_heap_inuse_bytes Number of heap bytes that are in use.
# TYPE go_memstats_heap_inuse_bytes gauge
go_memstats_heap_inuse_bytes 7.421952e+06
# HELP go_memstats_heap_objects Number of allocated objects.
# TYPE go_memstats_heap_objects gauge
go_memstats_heap_objects 29154
# HELP go_memstats_heap_released_bytes Number of heap bytes released to OS.
# TYPE go_memstats_heap_released_bytes gauge
go_memstats_heap_released_bytes 5.685248e+07
# HELP go_memstats_heap_sys_bytes Number of heap bytes obtained from system.
# TYPE go_memstats_heap_sys_bytes gauge
go_memstats_heap_sys_bytes 6.619136e+07
# HELP go_memstats_last_gc_time_seconds Number of seconds since 1970 of last garbage collection.
# TYPE go_memstats_last_gc_time_seconds gauge
go_memstats_last_gc_time_seconds 1.5878100098625474e+09
# HELP go_memstats_lookups_total Total number of pointer lookups.
# TYPE go_memstats_lookups_total counter
go_memstats_lookups_total 0
# HELP go_memstats_mallocs_total Total number of mallocs.
# TYPE go_memstats_mallocs_total counter
go_memstats_mallocs_total 509184
# HELP go_memstats_mcache_inuse_bytes Number of bytes in use by mcache structures.
# TYPE go_memstats_mcache_inuse_bytes gauge
go_memstats_mcache_inuse_bytes 3472
# HELP go_memstats_mcache_sys_bytes Number of bytes used for mcache structures obtained from system.
# TYPE go_memstats_mcache_sys_bytes gauge
go_memstats_mcache_sys_bytes 16384
# HELP go_memstats_mspan_inuse_bytes Number of bytes in use by mspan structures.
# TYPE go_memstats_mspan_inuse_bytes gauge
go_memstats_mspan_inuse_bytes 116960
# HELP go_memstats_mspan_sys_bytes Number of bytes used for mspan structures obtained from system.
# TYPE go_memstats_mspan_sys_bytes gauge
go_memstats_mspan_sys_bytes 147456
# HELP go_memstats_next_gc_bytes Number of heap bytes when next garbage collection will take place.
# TYPE go_memstats_next_gc_bytes gauge
go_memstats_next_gc_bytes 7.990544e+06
# HELP go_memstats_other_sys_bytes Number of bytes used for other system allocations.
# TYPE go_memstats_other_sys_bytes gauge
go_memstats_other_sys_bytes 637044
# HELP go_memstats_stack_inuse_bytes Number of bytes in use by the stack allocator.
# TYPE go_memstats_stack_inuse_bytes gauge
go_memstats_stack_inuse_bytes 917504
# HELP go_memstats_stack_sys_bytes Number of bytes obtained from system for stack allocator.
# TYPE go_memstats_stack_sys_bytes gauge
go_memstats_stack_sys_bytes 917504
# HELP go_memstats_sys_bytes Number of bytes obtained from system.
# TYPE go_memstats_sys_bytes gauge
go_memstats_sys_bytes 7.2958208e+07
# HELP go_threads Number of OS threads created.
# TYPE go_threads gauge
go_threads 10
# HELP ory_oathkeeper_request_duration_in_seconds Time spent serving requests.
# TYPE ory_oathkeeper_request_duration_in_seconds histogram
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.005"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.01"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.025"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.05"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.1"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.25"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="0.5"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="1"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="2.5"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="5"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="10"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200",le="+Inf"} 7
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200"} 0.0013910569999999998
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200"} 7
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.005"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.01"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.025"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.05"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.1"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.25"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="0.5"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="1"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="2.5"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="5"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="10"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200",le="+Inf"} 56
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200"} 0.009497334000000001
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200"} 56
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.005"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.01"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.025"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.05"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.1"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.25"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="0.5"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="1"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="2.5"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="5"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="10"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200",le="+Inf"} 2
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200"} 0.000384277
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200"} 2
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.005"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.01"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.025"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.05"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.1"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.25"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="0.5"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="1"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="2.5"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="5"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="10"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200",le="+Inf"} 168
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200"} 0.027453962999999994
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200"} 168
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.005"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.01"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.025"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.05"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.1"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.25"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="0.5"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="1"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="2.5"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="5"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="10"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200",le="+Inf"} 1
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200"} 0.000716618
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200"} 1
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.005"} 3
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.01"} 3
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.025"} 3
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.05"} 3
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.1"} 3
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.25"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="0.5"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="1"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="2.5"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="5"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="10"} 4
ory_oathkeeper_request_duration_in_seconds_bucket{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200",le="+Inf"} 4
ory_oathkeeper_request_duration_in_seconds_sum{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200"} 0.16967901900000001
ory_oathkeeper_request_duration_in_seconds_count{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200"} 4
# HELP ory_oathkeeper_request_total Total number of requests
# TYPE ory_oathkeeper_request_total counter
ory_oathkeeper_request_total{method="GET",request="/decisions/",service="oathkeeper-api",status_code="200"} 7
ory_oathkeeper_request_total{method="GET",request="/decisions/api/endpoint-A/",service="oathkeeper-api",status_code="200"} 56
ory_oathkeeper_request_total{method="GET",request="/decisions/api/endpoint-A/resource-D",service="oathkeeper-api",status_code="200"} 2
ory_oathkeeper_request_total{method="GET",request="/decisions/api/endpoint-A/resource-A",service="oathkeeper-api",status_code="200"} 168
ory_oathkeeper_request_total{method="GET",request="/decisions/api/endpoint-B/resource-B",service="oathkeeper-api",status_code="200"} 1
ory_oathkeeper_request_total{method="GET",request="/decisions/api/endpoint-B/resource-C",service="oathkeeper-api",status_code="200"} 4
# HELP process_cpu_seconds_total Total user and system CPU time spent in seconds.
# TYPE process_cpu_seconds_total counter
process_cpu_seconds_total 0.59
# HELP process_max_fds Maximum number of open file descriptors.
# TYPE process_max_fds gauge
process_max_fds 65536
# HELP process_open_fds Number of open file descriptors.
# TYPE process_open_fds gauge
process_open_fds 21
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 2.727936e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.58780960309e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 7.3895936e+08
# HELP process_virtual_memory_max_bytes Maximum amount of virtual memory available in bytes.
# TYPE process_virtual_memory_max_bytes gauge
process_virtual_memory_max_bytes -1
```

**Comment by aeneasr** [comment]
> So I'm not sure what you mean regarding the config but I've build the version from the PR locally and tested it by curling the scraping endpoint which looks good:

Yes, but you can't configure the host or the port on which this listens on. This can't be hardcoded!

File: cmd/server/server.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply cmd/server/server.go:207]: Sorry, I meant API endpoint!

File: metrics/middleware.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	return m
+}
+
+func (m *Middleware) ServeHTTP(rw http.ResponseWriter, r *http.Request, next http.HandlerFunc) {
+	start := m.clock.Now()
+	next(rw, r)
+	latency := m.clock.Since(start)
+	res := rw.(negroni.ResponseWriter)
+
+	if _, silent := m.silencePaths[r.URL.Path]; !silent {
```
[the remark this answers is not in the document]
Comment (ecktom) [reply metrics/middleware.go:67]: This is actually taken from your reqlog middleware and works in the same way https://github.com/ory/x/blob/4ac658cea58db29b417026d172fbfb6b567d1c9e/reqlog/middleware.go#L115

[the remark this answers is not in the document]
Comment (aeneasr) [reply metrics/middleware.go:67]: Right, should be fine then :)

File: cmd/server/server.go [hunk 3]
```diff
@@ -189,18 +190,35 @@ func RunServe(version, build, date string) func(cmd *cobra.Command, args []strin
...
+		}
+
+		promPath := d.Configuration().PrometheusMetricsPath()
+		promAddr := d.Configuration().PrometheusServeAddress()
+		go func() {
+			// Expose the registered metrics via HTTP.
+			httpServer := &http.Server{Handler: promhttp.HandlerFor(prometheusRepo.Registry, promhttp.HandlerOpts{}), Addr: promAddr}
+			http.Handle(promPath, promhttp.Handler())
+			logger.Infof("Proemtheus listening on %s...", promAddr)
+			if err := httpServer.ListenAndServe(); err != nil {
```
[the remark this answers is not in the document]
Comment (ecktom) [reply cmd/server/server.go]: done ;)

File: cmd/server/server.go [hunk 4]
```diff
@@ -189,18 +190,35 @@ func RunServe(version, build, date string) func(cmd *cobra.Command, args []strin
...
 		}
 
+		prometheusRepo, err := metrics.NewPrometheusRepository(logger)
+		if err != nil {
+			logger.Warnf("cannot create proemtehus repo: %+v", err)
+		}
+
+		promPath := d.Configuration().PrometheusMetricsPath()
+		promAddr := d.Configuration().PrometheusServeAddress()
+		go func() {
```
[the remark this answers is not in the document]
Comment (ecktom) [reply cmd/server/server.go]: done ;)

File: cmd/server/server.go [hunk 5]
```diff
@@ -189,18 +190,35 @@ func RunServe(version, build, date string) func(cmd *cobra.Command, args []strin
...
+			// Expose the registered metrics via HTTP.
+			httpServer := &http.Server{Handler: promhttp.HandlerFor(prometheusRepo.Registry, promhttp.HandlerOpts{}), Addr: promAddr}
```
[the remark this answers is not in the document]
Comment (ecktom) [reply cmd/server/server.go]: yeah right, I removed the returning error is it did not make sense anyways...

File: metrics/prometheus.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	// RequestTotal provides the total number of requests
+	RequestTotal = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "ory_oathkeeper_request_total",
+			Help: "Total number of requests",
+		},
+		[]string{"service", "method", "request", "status_code"},
+	)
+	// HistogramRequestDuration provides the duration of requests
+	HistogramRequestDuration = prometheus.NewHistogramVec(
```
[the remark this answers is not in the document]
Comment (ecktom) [reply metrics/prometheus.go:22]: Should follow https://prometheus.io/docs/practices/naming/

File: metrics/prometheus.go [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+		prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}),
+		RequestTotal,
+		HistogramRequestDuration,
+	}
+
+	r := prometheus.NewRegistry()
+
+	for _, metric := range m {
+		if err := r.Register(metric); err != nil {
+			logger.Warnf("register prometheus metrics: %+v", err)
```
[the remark this answers is not in the document]
Comment (ecktom) [reply metrics/prometheus.go:60]: Actually I don't think it worth a fatal but we can do this if you want

File: metrics/prometheus.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+type PrometheusRepository struct {
+	logger                 log.FieldLogger
+	requestDurationObserve RequestDurationObserve
+	updateRequest          UpdateRequest
+	Registry               *prometheus.Registry
+	metrics                []prometheus.Collector
+}
+
+// NewPrometheusRepository creates a new prometheus repository with the given settings
+func NewPrometheusRepository(logger log.FieldLogger) (*PrometheusRepository, error) {
```
[the remark this answers is not in the document]
Comment (ecktom) [reply metrics/prometheus.go]: fixed

**Comment by aeneasr** [comment]
If you merge with master, the weird failing test cases (flaky) will stop to happen :)

File: metrics/prometheus.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply metrics/prometheus.go:60]: Ok, makes sense! Let's make the error explicit in the logs then :)

```suggestion
			logger.WithError(err).Error("Unable to register prometheus metric.")
```

File: metrics/prometheus.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (aeneasr) [reply metrics/prometheus.go:22]: Perfect!

**Comment by ecktom** [comment]
Oh yeah, I've missed that comment :D Will add some words about this once it's running as I would also setup a Grafana board for us internally which I surely can share afterwards.
Should we test it e2e with the changes from today?

**Comment by aeneasr** [comment]
If you have ideas how to e2e test it that would be awesome!

**Comment by ecktom** [comment]
A dirty one... ;) Put it on our staging ENV, didn't crash or log any error and provided metrics on :9000/metrics as expected. So I guess it's good to go ;)

**Comment by aeneasr** [comment]
Ok nice, I thought you meant automated e2e test :D

Do you need help with the docs?

**Comment by ecktom** [comment]
Will come up with something next week, should be OK.

**Comment by aeneasr** [comment]
Awesome, thanks!

**Comment by aeneasr** [comment]
Thank you for your hard work!

**Pull request merged**